[WIP] build-script-impl: Partial implementation for a "quick" xctoolchain

jrose-apple opened the pull request:
Building an xctoolchain for use in Xcode is hard. utils/build-toolchain makes it easy, but it's not so great for actively working on swiftc because it always does a clean build and makes copies of the build products. Add support to build-script-impl (hopefully to be exposed in a pretty way elsewhere) for building a fake xctoolchain out of symlinks.

Only swiftc/SourceKit and LLDB are supported right now; someone else can do the work for playgrounds.

To play with this, try these commands with build-script:
```
  --darwin-toolchain-build-symlink
  --install-destdir=$PWD
  --toolchain-prefix=test.xctoolchain
  --darwin_toolchain_bundle_identifier=org.swift.MASTER
  --darwin-toolchain-display-name "Swift master"
```

jrose-apple commented:
Hey @modocache I finally did it. ;-)

I don't really want to own this. I just want to implement the bottom piece of it so someone else can take it and make it pretty. I don't care what any of it's named either.

zbowling commented:
![NEAT](https://media.giphy.com/media/x5kvlARMAkxri/giphy.gif)

modocache commented on utils/build-script-impl:
```diff
@@ -3384,6 +3385,49 @@ for host in "${ALL_HOSTS[@]}"; do
+    call ${PLISTBUDDY_BIN} -c "Add OverrideBuildSettings:SWIFT_LINK_OBJC_RUNTIME string 'YES'" "${DARWIN_TOOLCHAIN_INFO_PLIST}"
+
+    call mkdir -p "${host_install_destdir}${TOOLCHAIN_PREFIX}/usr/bin"
+    call ln -sf "$(build_directory_bin ${host} swift)/bin/"{swift,swiftc,swift-demangle,swift-stdlib-tool} "${host_install_destdir}${TOOLCHAIN_PREFIX}/usr/bin"
```
When I run this locally, I seem to be getting the path `/Users/bgesiak/Source/apple/build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/bin/swiftc` -- note the extra `/bin`.  The following works for me:

```diff
diff --git a/utils/build-script-impl b/utils/build-script-impl
index 512aeff8f4..4a0e6e6a28 100755
--- a/utils/build-script-impl
+++ b/utils/build-script-impl
@@ -3415,7 +3415,7 @@ function build_symlink_darwin_toolchain() {
     call ${PLISTBUDDY_BIN} -c "Add OverrideBuildSettings:SWIFT_LINK_OBJC_RUNTIME string 'YES'" "${DARWIN_TOOLCHAIN_INFO_PLIST}"

     call mkdir -p "${host_install_destdir}${TOOLCHAIN_PREFIX}/usr/bin"
-    call ln -sf "$(build_directory_bin ${host} swift)/bin/"{swift,swiftc,swift-demangle,swift-stdlib-tool} "${host_install_destdir}${TOOLCHAIN_PREFIX}/usr/bin"
+    call ln -sf "$(build_directory_bin ${host} swift)/"{swift,swiftc,swift-demangle,swift-stdlib-tool} "${host_install_destdir}${TOOLCHAIN_PREFIX}/usr/bin"
     call mkdir -p "${host_install_destdir}${TOOLCHAIN_PREFIX}/usr/lib"
     call ln -sf "$(build_directory_bin ${host} swift)/../lib/"{swift,swift_static,libswiftDemangle.dylib,sourcekitd.framework} "${host_install_destdir}${TOOLCHAIN_PREFIX}/usr/lib"
```

jrose-apple replied:
Whoops, I fixed this for LLDB but not for Swift.

modocache reviewed (changes requested):
Oh yes, fantastic!

+cc @kastiglione, since I believe he's expressed interest in this before.

I pulled this down and tried it myself. I had one little error that I noted below (which is why I'm requesting changes), but otherwise I was amazed at how quickly I was able to produce a working toolchain.

I think one potential improvement, for the person who would want to "own" this, would maybe be to script-ify the `PlistBuddy` and symlink portion. My reasoning here is that, for people who are OK with some really sharp edges, it'd be nice to run that part without invoking `build-script` or `build-script-impl`.

Maybe even better still would be moving this to CMake. That would allow `cmake --build ${BUILD_DIR_PATH} --target darwin-toolchain`, which would be very handy, because it would allow a symlink toolchain to be generated, but without the overhead of running CMake configuration.

I think one benefit of turning this into a script, as opposed to CMake, would be the ability to specify an `--install-destdir=` to a path that I'd need `sudo` to write to -- for example, if I wanted to put this directly into my toolchains directory `/Library/Developer/Toolchains/`. Right now, I run this script with the command you described, `utils/build-script --release -- --darwin-toolchain-build-symlink --install-destdir=/path/to/swift-darwin-toolchain --toolchain-prefix=test.xctoolchain --darwin_toolchain_bundle_identifier=org.swift.MASTER --darwin-toolchain-display-name "Swift master"`, and then run `sudo cp -r /path/to/swift-darwin-toolchain/swift-darwin-toolchain/test.xctoolchain /Library/Developer/Toolchains/test.xctoolchain`. Maybe I'm being paranoid, but for some reason I'm worried to run `sudo utils/build-script` or `sudo cmake --build`, and so building and copying a toolchain to the correct location requires two steps for me.

Anyway this is so, so great. Thanks for doing this!

jrose-apple commented:
I don't love having it in build-script-impl, but it seems wrong to put it in CMake at the moment when it also deals with LLDB. It can't go in a standalone script either because it needs to know where the build products are going to end up.

jrose-apple committed:
build-script-impl: Partial implementation for a "quick" xctoolchain

Building an xctoolchain for use in Xcode is hard.
utils/build-toolchain makes it easy, but it's not so great for
actively working on swiftc. Add support to build-script-impl
(hopefully to be exposed in a pretty way elsewhere) for building a
fake xctoolchain out of symlinks.

Only swiftc/SourceKit and LLDB are supported right now; someone else
can do the work for playgrounds.

To play with this, try these commands with build-script:
  --darwin-toolchain-build-symlink
  --install-destdir=$PWD
  --toolchain-prefix=test.xctoolchain
  --darwin_toolchain_bundle_identifier=org.swift.MASTER
  --darwin-toolchain-display-name "Swift master"

modocache reviewed (approved):
Oh, I see. For the standalone script idea, couldn't the script take the path to the build product directory as an argument...?

In any case, thanks for the fix, this looks great!

jrose-apple commented:
It's not just the build product directory (`build/$BUILD_SUBDIR`); there's also an extra level of subdirectory for Xcode builds (including all LLDB builds) based on the build mode. You could use the swift_build_support stuff to parse that out from command-line arguments or a preset, though.

benlangmuir commented:
Neat, this can hopefully obsolete a script that I'm using 😃 I have a few more paths that I'm handling that come from llvm/clang:
```
bin/clang
lib/libclang.dylib
lib/clang
lib/libIndexStore.dylib
include/indexstore ## comes from source directory, not build
``` 
Any problem including these as well? I'm happy to add them myself next time I want this functionality.

jrose-apple commented:
I was unsure if we wanted to handle those, since we don't include them in the downloadable toolchains. Then again, it's clearly necessary for people working on SourceKit on cross-language-boundary features.

benlangmuir commented:
Hmm, I guess we should do our best to match whatever `build-toolchain` does.  We could add a separate build option that includes the additional products that I need.

futurejones commented:
Can one of the admins verify this patch?

CodaFi commented:
@jrose-apple Now that we have the ability to have CI build us toolchains, do you think we should keep this around?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
